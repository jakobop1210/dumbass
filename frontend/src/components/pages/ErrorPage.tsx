import Button from '../atoms/Button'
import Breadcrumb from '../atoms/Breadcrumb'

/**
 * NotFoundPage component - A user-friendly error page for routing failures
 *
 * This component is displayed when the user tries to navigate to a route that does not exist.
 * It provides an error message and a button to navigate back to the homepage.
 */

export type ErrorPageType = {
    title: string
    subTitle: string
    description: string
    buttonText: string
    buttonFunction: () => void
}

export const ErrorPage = (props: ErrorPageType) => {
    return (
        <main>
            <Breadcrumb
                items={[
                    {
                        name: 'Unknown Page'
                    }
                ]}
            />
            <section className='flex items-center pt-10 px-5 sm:pt-32 sm:px-10 flex-col text-gray-500 text-center'>
                <header>
                    <h1 className='text-7xl'>{props.title}</h1>
                    <h2 className='text-3xl my-4'>{props.subTitle}</h2>
                </header>

                <section className='max-w-xl mb-10'>
                    <p>{props.description}</p>
                </section>
                <section className='w-1/6 min-w-fit'>
                    <Button
                        title={props.buttonText}
                        type='button'
                        onClick={() => props.buttonFunction()}
                    />
                </section>
            </section>
        </main>
    )
}
