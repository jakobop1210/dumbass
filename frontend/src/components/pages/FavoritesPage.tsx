import { useSelector } from 'react-redux'
import { RootState } from '../../redux/store'
import { Artist, Song } from '../../lib/types'
import { ClipLoader } from 'react-spinners'
import Breadcrumb from '../atoms/Breadcrumb'
import CardView from '../organisms/CardView'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import { useFavorites } from '../../hooks/useFavorites'

export default function FavoritesPage() {
    const username = useSelector((state: RootState) => state.user.username)

    const { dataSongs, loading, error, dataArtists } = useFavorites(username)

    // Processing and mapping data for Card components
    const artists: Artist[] = dataArtists?.getArtistsByIds || []
    const songs: Song[] = dataSongs?.getSongsByIds || []

    const artistCardData = artists.map(
        (artist: Artist): ArtistCardProps => ({
            cardType: 'artist',
            id: artist.id,
            title: artist.name,
            alternateNames: artist.alternate_names,
            imageUrl: artist.image_url,
            rating: artist.average_rating,
            numOfRatings: artist.number_of_ratings
        })
    )

    const songCardData = songs.map(
        (song: Song): SongCardProps => ({
            cardType: 'song',
            id: song.id,
            title: song.title,
            artist: song.artist_names,
            imageUrl: song.header_image_url,
            rating: song.average_rating,
            numOfRatings: song.number_of_ratings,
            releaseDate: song.release_date
        })
    )

    if (!username) {
        return (
            <div className='h-96 flex items-center justify-center text-blueGray italic'>
                <h1>You need to be logged in to add favorites...</h1>
            </div>
        )
    }

    return (
        <main className='w-full flex flex-col'>
            <Breadcrumb items={[{ name: 'Favorites' }]} />
            <section className='w-full flex flex-col justify-center items-center mt-5'>
                {loading ? (
                    <div className='h-96 flex items-center'>
                        <ClipLoader color={'#8fc0a9'} size={100} />
                    </div>
                ) : error ? (
                    <div className='h-96 flex items-center'>
                        <h1 className='text-blueGray text-2xl'>
                            Error loading data :/
                        </h1>
                    </div>
                ) : (
                    <>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Songs'
                                cardData={songCardData}
                                customErrorMessage='You have no favorite songs...'
                            />
                        </section>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Artists'
                                cardData={artistCardData}
                                customErrorMessage='You have no favorite artists...'
                            />
                        </section>
                    </>
                )}
            </section>
        </main>
    )
}
