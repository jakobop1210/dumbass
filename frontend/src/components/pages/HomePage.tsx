import { useState } from 'react'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import CardView from '../organisms/CardView'
import CommonSearchBar from '../molecules/CommonSearchBar'
import { GET_TOP_ARTISTS } from '../../graphql/queries/artistQueries'
import { GET_TOP_SONGS } from '../../graphql/queries/songQueries'
import { useQuery } from '@apollo/client'
import { customToast } from '../../lib/utils'
import { ClipLoader } from 'react-spinners'

// Define a type for an artist
type Artist = {
    id: string
    name: string
    image_url: string
    alternate_names: string[]
    average_rating: number
    number_of_ratings: number
}

// Define a type for a song
type Song = {
    id: string
    title: string
    artist_names: string
    header_image_url: string
    release_date: string
    primary_artist_id: number
    average_rating: number
    number_of_ratings: number
}

/**
 * @component ExplorePage
 *
 * `ExplorePage` is a React functional component that displays a page where users
 * can explore various artists and songs using a search bar and view them via card layouts.
 */
export default function HomePage() {
    const [filter, setFilter] = useState('Song')
    const limitData = 12

    // Define GraphQL queries using the useQuery hook
    const {
        loading: loadingArtists,
        error: errorArtists,
        data: dataArtists
    } = useQuery(GET_TOP_ARTISTS, {
        variables: { limit: limitData }
    })
    const {
        loading: loadingSongs,
        error: errorSongs,
        data: dataSongs
    } = useQuery(GET_TOP_SONGS, {
        variables: { limit: limitData }
    })

    if (errorArtists || errorSongs) {
        console.log(
            `Error: ${JSON.stringify(errorArtists || errorSongs, null, 2)}`
        )

        customToast('error', 'Error', 'Could not load data')
    }

    // Combine and display data
    const artists: Artist[] = dataArtists?.getTopArtists || []
    const songs: Song[] = dataSongs?.getTopSongs || []

    const artistCardData: ArtistCardProps[] = artists.map((artist: Artist) => ({
        cardType: 'artist',
        id: artist.id,
        title: artist.name,
        alternateNames: artist.alternate_names,
        imageUrl: artist.image_url,
        rating: artist.average_rating,
        numOfRatings: artist.number_of_ratings
    }))

    const songCardData: SongCardProps[] = songs.map((song: Song) => ({
        cardType: 'song',
        id: song.id,
        title: song.title,
        artist: song.artist_names,
        imageUrl: song.header_image_url,
        rating: song.average_rating,
        numOfRatings: song.number_of_ratings,
        releaseDate: song.release_date
    }))

    return (
        <main className='flex flex-col items-center justify-center w-screen'>
            <CommonSearchBar
                className='w-4/5 mt-10 drop-shadow mb-10'
                filterOptions={['Song', 'Artist']}
                selectedFilter={filter}
                onFilterChange={(newFilter) => setFilter(newFilter)}
            />
            {loadingArtists || loadingSongs ? (
                <div className='h-96 flex items-center'>
                    <ClipLoader color={'#8fc0a9'} size={100} />
                </div>
            ) : (
                <>
                    {errorArtists || errorSongs ? (
                        <div className='h-96 flex items-center'>
                            <h1 className='text-blueGray text-2xl'>
                                Error loading data :/
                            </h1>
                        </div>
                    ) : (
                        <>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Songs'
                                    cardData={songCardData}
                                />
                            </section>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Artists'
                                    cardData={artistCardData}
                                />
                            </section>
                        </>
                    )}
                </>
            )}
        </main>
    )
}
