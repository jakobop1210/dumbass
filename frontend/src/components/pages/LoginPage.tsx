import { FaMusic } from 'react-icons/fa'
import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import InputField from '../atoms/InputField'
import Button from '../atoms/Button'
import { customToast } from '../../lib/utils'
import { LOGIN_USER } from '../../graphql/mutations/userMutations'
import { useMutation } from '@apollo/client'
import { useDispatch } from 'react-redux';
import { setUserLogin, setUserName } from '../../redux/actions/userActions';

/**
 * LoginPage component - Used for user authentication
 *
 * Allows a user to log in by providing a username and a password. After successful login, navigation to the homepage is performed and a successful login message is shown. If the login fails, an error message is displayed.
 *
 * @param {LoginPageProps} props - Properties passed down from parent component. Optionally includes `setUser`.
 */
export default function LoginPage(): JSX.Element {
    const navigate = useNavigate()
    const dispatch = useDispatch()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')
    const [loginUserMutation] = useMutation(LOGIN_USER)

    /**
     * loginUser - Event handler for form submission.
     *
     * Validates provided username and password. If credentials match (username: 'guest', password: 'guest'), navigates to the homepage and optionally invokes the `setUser` function if provided.
     * Shows a toast notification for successful login or failure.
     *
     * @param {React.FormEvent} e - Event object related to the form submission.
     */
    async function loginUser(e: React.FormEvent) {
        e.preventDefault()

        try {
            const result = await loginUserMutation({
                variables: { username, password }
            })
            const { data } = result

            if (data.loginUser) {
                customToast('success', 'Successfully logged in')
                dispatch(setUserLogin()); // Dispatch Redux action to set user as logged in
                dispatch(setUserName(username)); // Dispatch Redux action to set username
               
                // Store username and login status in local storage
                localStorage.setItem('username', username);
                localStorage.setItem('isLoggedIn', 'true');

                navigate('/')
            } else {
                customToast('error', 'An error occurred while logging in')
            }
        } catch (error) {
            console.log(JSON.stringify(error, null, 2))
            customToast('error', 'Wrong password or username')
        }
    }

    return (
        <main className='w-screen flex justify-center'>
            <section className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <header className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <div className='flex items-center justify-center'>
                        <FaMusic size={60} color='#8FC0A9' />
                    </div>
                    <h2 className='mt-10 text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Sign in to your account
                    </h2>
                </header>

                <div className='mt-10 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={loginUser}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <Button title={'Sign in'} type={'submit'} />
                    </form>

                    <footer className='mt-10 text-center text-sm text-blueGray'>
                        Don't have an account?
                        <button
                            onClick={() => navigate('/register')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Register here
                        </button>
                    </footer>
                </div>
            </section>
        </main>
    )
}
