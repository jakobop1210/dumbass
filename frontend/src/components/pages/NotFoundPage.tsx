import { useNavigate } from 'react-router-dom'
import Button from '../atoms/Button'

/**
 * NotFoundPage component - A user-friendly error page for routing failures
 *
 * This component is displayed when the user tries to navigate to a route that does not exist.
 * It provides an error message and a button to navigate back to the homepage.
 */
const NotFoundPage = () => {
    const navigate = useNavigate()

    return (
        <main className='flex items-center pt-10 px-5 sm:pt-32 sm:px-10 flex-col text-gray-500 text-center'>
            <header>
                <h1 className='text-7xl'>Oops!</h1>
                <h2 className='text-3xl my-4'>404 - Page Not Found</h2>
            </header>
            
            <section className='max-w-xl mb-10'>
                <p>
                    The page you are looking for does not exist. How you got here is
                    a mystery, but you can click the button below to go back to the
                    homepage.
                </p>
            </section>
            <section className='w-1/6'>
                <Button
                    title='Go to homescreen'
                    type='button'
                    onClick={() => navigate('/')}
                />
            </section>
        </main>
    )
}

export default NotFoundPage
