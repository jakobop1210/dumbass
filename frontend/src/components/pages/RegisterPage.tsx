import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import InputField from '../atoms/InputField'
import Button from '../atoms/Button'
import { customToast } from '../../lib/utils'
import { ADD_USER } from '../../graphql/mutations/userMutations'
import { useMutation } from '@apollo/client'

/**
 * `RegisterPage` Component.
 *
 * This component presents a registration form for creating a new user account. 
 * It includes input fields for username, password, and password confirmation, along with a submit button.
 * The form performs validation to ensure that the entered passwords match.
 *
 * Upon form submission, a GraphQL mutation is used to attempt creating a new user. 
 * Success or failure feedback is provided to the user through custom toast messages.
 * On successful account creation, the user is navigated to the login page.
 * In case of errors such as a username conflict, an appropriate error message is displayed.
 *
 * The page also provides a link to the login page for users who already have an account.
 *
 * @returns {JSX.Element} The rendered registration page with a form for creating a new user account.
 */
export default function RegisterPage(): JSX.Element {
    const navigate = useNavigate()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')
    const [confirmPassword, setConfirmPassword] = useState('')
    const [addUser] = useMutation(ADD_USER)

    /**
     * Handles the registration form submission.
     * Validates the form data and uses the `ADD_USER` mutation for creating a new user.
     *
     * @param {React.FormEvent} e - The form event.
     */
    async function registerAccount(e: React.FormEvent) {
        e.preventDefault()
        if (password !== confirmPassword) {
            customToast('error', 'The passwords does not match')
            return
        }
        try {
            const { data } = await addUser({
                variables: { username, password }
            })

            if (data.addUser) {
                customToast('success', 'User successfully created')
                navigate('/login')
            } else {
                customToast('error', 'Failed to create the user')
            }
        } catch (error) {
            customToast('error', 'Username is already taken')
        }
    }

    return (
        <main className='w-screen flex justify-center'>
            <section className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <header className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <h2 className='text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Create an account
                    </h2>
                </header>

                <div className='mt-6 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={registerAccount}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <InputField
                            id='confirmPassword'
                            type='password'
                            title='Confirm Password'
                            value={confirmPassword}
                            onChange={setConfirmPassword}
                            required
                        />
                        <Button title='Create Account' type='submit' />
                    </form>

                    <footer className='mt-10 text-center text-sm text-blueGray'>
                        Already have an account?
                        <button
                            onClick={() => navigate('/login')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Login here
                        </button>
                    </footer>
                </div>
            </section>
        </main>
    )
}
