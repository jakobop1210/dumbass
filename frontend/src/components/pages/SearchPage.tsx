import { useEffect, useState } from 'react'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import CardView from '../organisms/CardView'
import SearchBar from '../molecules/SearchBar'
import Breadcrumb from '../atoms/Breadcrumb'
import { useLocation } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import {
    COUNT_ARTISTS,
    GET_ARTISTS_ON_NAME
} from '../../graphql/queries/artistQueries'
import {
    COUNT_SONGS,
    GET_SONGS_ON_TITLE
} from '../../graphql/queries/songQueries'
import Pagination from '../molecules/Pagination'
import { Artist, Song } from '../../lib/types'
import { useSearchParams } from 'react-router-dom'
import Dropdown from '../atoms/Dropdown'
import { ClipLoader } from 'react-spinners'

type CardProps = ArtistCardProps | SongCardProps

/**
 * SearchPage component to render and handle search functionality,
 * filtering and sorting for artists and songs.
 */
function SearchPage() {
    const location = useLocation()
    const queryParams = new URLSearchParams(location.search)

    const term = queryParams.get('term')
    const filter = queryParams.get('filter')
    const sort = queryParams.get('sort')
    const [searchParams, setSearchParams] = useSearchParams()

    const validFilters = ['song', 'artist']
    const defaultFilter = 'song'
    const validSort = ['relevance', 'rating', 'alphabetical']
    const defaultSort = 'relevance'

    const filterFromURL = searchParams.get('filter') || defaultFilter
    const sortFromURL = searchParams.get('sort') || defaultSort

    const { data: totalSongsData } = useQuery(COUNT_SONGS, {
        variables: { title: term },
        skip: filter !== 'song'
    })

    const { data: totalArtistsData } = useQuery(COUNT_ARTISTS, {
        variables: { name: term },
        skip: filter !== 'artist'
    })

    const [selectedFilter, setSelectedFilter] = useState(
        validFilters.includes(filterFromURL) ? filterFromURL : defaultFilter
    )
    const [selectedSort, setSelectedSorting] = useState(
        validSort.includes(sortFromURL) ? sortFromURL : defaultSort
    )

    useEffect(() => {
        const newSearchParams = new URLSearchParams(searchParams)
        if (searchParams.get('filter') !== selectedFilter) {
            newSearchParams.set('filter', selectedFilter)
        }
        if (searchParams.get('sort') !== selectedSort) {
            newSearchParams.set('sort', selectedSort)
        }
        setSearchParams(newSearchParams, { replace: true })
    }, [selectedFilter, selectedSort, searchParams, setSearchParams])

    const [data, setData] = useState<CardProps[]>([])
    const [currentPage, setCurrentPage] = useState(1)
    const [totalPages, setTotalPages] = useState(0)
    const itemsPerPage = 12

    const { data: artistsData, loading: artistsLoading } = useQuery(
        GET_ARTISTS_ON_NAME,
        {
            variables: {
                name: term,
                limit: itemsPerPage,
                sort: sort,
                page: currentPage
            },
            skip: filter !== 'artist',
            fetchPolicy: 'cache-and-network'
        }
    )

    const { data: songsData, loading: songsLoading } = useQuery(
        GET_SONGS_ON_TITLE,
        {
            variables: {
                title: term,
                limit: itemsPerPage,
                sort: sort,
                page: currentPage
            },
            skip: filter !== 'song',
            fetchPolicy: 'cache-and-network'
        }
    )

    useEffect(() => {
        if (filter === 'artist' && artistsData) {
            const artistCardData: ArtistCardProps[] =
                artistsData.getArtistsOnName.map((artist: Artist) => ({
                    cardType: 'artist',
                    id: artist.id,
                    title: artist.name,
                    alternateNames: artist.alternate_names,
                    imageUrl: artist.image_url,
                    rating: artist.average_rating,
                    numOfRatings: artist.number_of_ratings
                }))
            setData(artistCardData)
        } else if (filter === 'song' && songsData) {
            const songCards: SongCardProps[] = songsData.getSongsOnTitle.map(
                (song: Song) => ({
                    cardType: 'song',
                    id: song.id,
                    title: song.title,
                    artist: song.artist_names,
                    imageUrl: song.header_image_url,
                    rating: song.average_rating,
                    numOfRatings: song.number_of_ratings,
                    releaseDate: song.release_date
                })
            )
            setData(songCards)
        }
    }, [filter, artistsData, songsData])

    useEffect(() => {
        let totalItems = 0
        if (filter === 'artist' && totalArtistsData) {
            totalItems = totalArtistsData.countArtists
        } else if (filter === 'song' && totalSongsData) {
            totalItems = totalSongsData.countSongs
        }
        setTotalPages(Math.ceil(totalItems / itemsPerPage))
    }, [totalArtistsData, totalSongsData, itemsPerPage, filter])

    useEffect(() => {
        setCurrentPage(1)
    }, [term, selectedFilter, selectedSort])

    const isTermProvided = term !== null && term.trim() !== ''

    if (!isTermProvided) {
        return (
            <main className='w-full'>
                <Breadcrumb items={[{ name: 'Search' }]} />
                <header className='flex justify-center'>
                    <SearchBar
                        filterOptions={validFilters}
                        selectedFilter={selectedFilter}
                        selectedSort={selectedSort}
                        onFilterChange={(newFilter) =>
                            setSelectedFilter(newFilter)
                        }
                    />
                </header>
                <section className='flex text-xl justify-center w-full text-gray-500'>
                    Please enter a search term...
                </section>
            </main>
        )
    }

    return (
        <main className='w-full'>
            <Breadcrumb
                items={[
                    {
                        name: 'Search'
                    }
                ]}
            />
            <header className='flex justify-center'>
                <SearchBar
                    filterOptions={validFilters}
                    selectedFilter={selectedFilter}
                    selectedSort={selectedSort}
                    onFilterChange={(newFilter) => setSelectedFilter(newFilter)}
                />
            </header>
            <section className='flex justify-center mt-5 mb-10'>
                {data.length > 0 && (
                    <Dropdown
                        selectedFilter={selectedSort}
                        filterOptions={validSort}
                        onFilterChange={(newSorting) =>
                            setSelectedSorting(newSorting)
                        }
                        outsideSearchBar
                        title='Sort by'
                    />
                )}
            </section>
            <section className='w-full flex flex-col justify-center items-center'>
                <section className='w-full flex justify-center'>
                    {artistsLoading || songsLoading ? (
                        <div className='absolute h-96 flex items-center'>
                            <ClipLoader color={'#8fc0a9'} size={100} />
                        </div>
                    ) : (
                        <CardView cardData={data} />
                    )}
                </section>
                <Pagination
                    onClickPrevious={() => setCurrentPage(currentPage - 1)}
                    onClickNext={() => setCurrentPage(currentPage + 1)}
                    currentPage={currentPage}
                    totalPages={totalPages}
                    isLoading={artistsLoading || songsLoading}
                />
            </section>
        </main>
    )
}

export default SearchPage
