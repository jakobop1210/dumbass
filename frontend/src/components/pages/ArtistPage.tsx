import { useNavigate, useParams } from 'react-router-dom'
import { InfoPageTemplate } from '../organisms/InfoPageTemplate'
import { useQuery } from '@apollo/client'
import { GET_ARTIST_BY_ID } from '../../graphql/queries/artistQueries'
import { faCircleInfo } from '@fortawesome/free-solid-svg-icons'
import { formatAlternateNames } from '../../lib/utils'
import { ErrorPage } from './ErrorPage'

/**
 * `ArtistPage` Component.
 *
 * This component fetches and displays detailed information about an artist using an `InfoPageTemplate`.
 * It queries the data for a specific artist by their ID using GraphQL and renders the `InfoPageTemplate` with this data.
 * The page displays the artist's name, alternate names, image, rating, and description. It includes a tab for additional artist information.
 *
 * If the data is not available or an error occurs, the component renders an `ErrorPage` component,
 * providing the user with a message and an option to navigate back to the homepage.
 *
 * @returns {JSX.Element} The rendered page with either the artist information or an error message.
 */
export const ArtistPage = () => {
    const { id = '' } = useParams()
    const navigate = useNavigate()

    // GraphQL query to fetch artist details by ID
    const { data, loading } = useQuery(GET_ARTIST_BY_ID, {
        variables: { id: parseInt(id) }
    })
    
    // Conditional rendering based on data availability or loading status
    if (data?.getArtistById || loading) {
        return (
            <InfoPageTemplate
                isLoading={loading}
                title={data?.getArtistById.name}
                subtitle={formatAlternateNames(
                    data?.getArtistById.alternate_names
                )}
                image={data?.getArtistById.image_url}
                averageRating={data?.getArtistById.average_rating}
                numOfRatings={data?.getArtistById.number_of_ratings}
                description={data?.getArtistById.description}
                tabs={[
                    {
                        title: 'Info',
                        icon: faCircleInfo
                    }
                ]}
                id={id}
                type='artist'
            />
        )
    } else {
        return (
            <ErrorPage
                title='Oops!'
                subTitle='Unknown error'
                description='Something wrong happened, you can try again or click the button below to go back to the homepage'
                buttonText='Go to homepage'
                buttonFunction={() => navigate('/')}
            />
        )
    }
}
