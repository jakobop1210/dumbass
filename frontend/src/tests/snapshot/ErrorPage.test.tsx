import { render } from '@testing-library/react';
import { ErrorPage } from '../../components/pages/ErrorPage';
import { expect, test, vi } from 'vitest';
import { BrowserRouter as Router } from 'react-router-dom';

test('ErrorPage snapshot', () => {
    const mockFunction = vi.fn();

    const props = {
        title: "Error Occurred",
        subTitle: "Page Not Found",
        description: "The page you are looking for might have been removed, had its name changed, or is temporarily unavailable.",
        buttonText: "Go Back",
        buttonFunction: mockFunction
    };

    const { asFragment } = render(
        <Router>
            <ErrorPage {...props} />
        </Router>
    );

    expect(asFragment()).toMatchSnapshot();
});