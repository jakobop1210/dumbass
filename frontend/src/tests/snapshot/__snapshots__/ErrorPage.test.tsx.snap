// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage snapshot 1`] = `
<DocumentFragment>
  <main>
    <div
      class="w-1/2 text-left absolute"
    >
      <nav
        aria-label="breadcrumb"
        class="w-max"
      >
        <ol
          class="flex w-full flex-wrap items-center rounded-md bg-opacity-60 py-2 px-4 "
        >
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <a
              class="opacity-60 duration-300 hover:text-green"
              href="/"
              id="breadcrumb-link-home"
            >
              <svg
                class="h-4 w-4"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
                />
              </svg>
            </a>
            <span
              class="pointer-events-none mx-2 select-none font-sans text-sm font-normal leading-normal antialiased"
            >
              /
            </span>
          </li>
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <div
              class="opacity-70"
            >
              <span>
                Unknown Page
              </span>
            </div>
          </li>
        </ol>
      </nav>
    </div>
    <section
      class="flex items-center pt-10 px-5 sm:pt-32 sm:px-10 flex-col text-gray-500 text-center"
    >
      <header>
        <h1
          class="text-7xl"
        >
          Error Occurred
        </h1>
        <h2
          class="text-3xl my-4"
        >
          Page Not Found
        </h2>
      </header>
      <section
        class="max-w-xl mb-10"
      >
        <p>
          The page you are looking for might have been removed, had its name changed, or is temporarily unavailable.
        </p>
      </section>
      <section
        class="w-1/6 min-w-fit"
      >
        <button
          class="flex w-full justify-center rounded-md px-3 py-1.5 text-sm font-semibold leading-8 text-white shadow-[0_4px_9px_-4px_#8FC0A9] undefined"
          type="button"
        >
          Go Back
        </button>
      </section>
    </section>
  </main>
</DocumentFragment>
`;
